### PR TITLE
Add useLoadScript and LoadScriptNext

### LINKED ISSUE #139: LoadScript used multiple times in app
I've already tried to discuss this [on the spectrum](https://spectrum.chat/react-google-maps/general/you-have-included-the-google-maps-javascript-api-multiple-times-on-this-page~99cc4656-9502-4fc4-9fa2-9cdccc8f44f7), but no response, so let's poke a bear in here :)

There is an obvious problem that `LoadScript` injects a script to header every time it is mounted. It's not keeping any information it was already loaded before. This is not only a problem of fetching the same thing multiple times, but Google API seems to have a problem with that and it causes weird issues like __`google` is not defined__

Furthermore, it shouldn't be just a boolean flag to remember it has been loaded, because it might be needed with a different set of props.

The react-google-maps are using `scriptjs` module and I have a feeling it actually tracks fetching the same resource multiple times and that's why it works.

So why you have opted not to use `scriptjs` here I wonder?

### PR DESCRIPTION
**Pull request by FredyC**
Fixes #139 

As noted in #139, instead of modifying current LoadScript I have created an alternative approach with Hooks and without that cleanup logic which I think was causing more issues than it helped.

I've added this build into our app and so far it works correctly and intend to push it to the production soon (yes, I am that bold) so we will see if it actually helps.

I've also decided to keep whole logic in the `useLoadScript` hook instead of a component to provide loading state directly instead of the callback. This is the way I am using it and works great. 

```tsx
export const BaseMap = observer<IBaseMapProps>(
  ({ model, children }) => {
    const { user } = useRoot()

    const { isLoaded } = useLoadScript({
      id: 'script-loader',
      version: 'weekly',
      googleMapsApiKey: appConfig.googleMapsApi!,
      libraries,
      language: user.language,
    })

    const renderMap = () => (
      <GoogleMap
        zoom={model.zoom}
        center={model.center}
        onCenterChanged={model.updatePosition}
        onZoomChanged={model.updateZoom}
        onClick={model.onClick}
        onLoad={model.setMapRef}
        onUnmount={() => model.setMapRef(null)}
        options={getMapOptions(isMobile)}
      >
        {children}
      </GoogleMap>
    )

    return isLoaded ? renderMap() : <Loading />
  },
)
```

For the convenience (and possible migration path), there is the `LoadScriptNext` component, but personally, I think it's not that useful anymore.

cc @JustFly1984 @uriklar @orta-sanz

### CODE REVIEW

File: packages/react-google-maps-api/src/useLoadScript.ts [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+        }
+        console.warn(`
+        There has been an Error with loading Google Maps API script, please check that you provided correct google API key to <LoadScript /> (${googleMapsApiKey})
+        Otherwise it is a Network issues.
+      `)
+        console.error(err)
+      })
+
+    return () => {
+      setLoadedIfMounted(false)
```
Comment (FredyC) [inline packages/react-google-maps-api/src/useLoadScript.ts]: I am thinking this might be actually a bad idea. It would be causing a switch back to the loading state on unmount and possibly some unwanted flicker.

File: packages/react-google-maps-api/src/useLoadScript.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    return () => {
+      isMounted.current = false
+      if (onUnmount) {
+        onUnmount()
+      }
+    }
+  }, [])
+
+  React.useEffect(() => {
+    if (isBrowser && preventGoogleFontsLoading) {
```
Comment (FredyC) [inline packages/react-google-maps-api/src/useLoadScript.ts]: I haven't fully understood why is this needed, so I kept it, but `react-google-maps` had nothing similar so I wonder what's the purpose of it?

Comment (JustFly1984) [reply packages/react-google-maps-api/src/useLoadScript.ts]: That is my own addition, cos I do not want to google maps to include google font Roboto. I already have Roboto fonts loaded with async in my app.

File: packages/react-google-maps-api/src/useLoadScript.ts (hunk 1 above)
Comment (JustFly1984) [reply packages/react-google-maps-api/src/useLoadScript.ts]: that effect is only running if id or url has changed

**Comment by JustFly1984** [comment]
Where did you import that `observer` function from?

**Comment by FredyC** [comment]
@JustFly1984 Nevermind, that's just MobX stuff, doesn't matter for the sake of the example :)

File: packages/react-google-maps-api/src/useLoadScript.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        There has been an Error with loading Google Maps API script, please check that you provided correct google API key to <LoadScript /> (${googleMapsApiKey})
+        Otherwise it is a Network issues.
+      `)
+        console.error(err)
+      })
+
+    return () => {
+      setLoadedIfMounted(false)
+    }
+  }, [id, url])
```
Comment (FredyC) [inline packages/react-google-maps-api/src/useLoadScript.ts]: I should probably include the `onLoad` and `onError` here as well, but it's a kinda bad idea to be re-injecting script if a parent is not having `useCallback` for these. So I think I will have to keep those in a ref otherwise it would see only first instances. Any idea of how to tackle it better?

File: packages/react-google-maps-api/src/useLoadScript.ts (hunk 1 above)
Comment (FredyC) [reply packages/react-google-maps-api/src/useLoadScript.ts]: Yea, but also on unmount of the component which would cause a short flicker back to a loading state.

File: packages/react-google-maps-api/src/useLoadScript.ts (hunk 3 above)
Comment (FredyC) [reply packages/react-google-maps-api/src/useLoadScript.ts]: Nevermind, figured out a better way with separate useEffect calls. Although it might be better to keep those callbacks out of the hook and use them only for a component variant.

**Comment by JustFly1984** [comment]
I think in example you should not generate `getMapOptions(isMobile)` on each function call, but use ```React.memo(getMapOptions(isMobile), [isMobile])```

**Comment by FredyC** [comment]
> I think in example you should not generate `getMapOptions(isMobile)` on each function call, but use `React.memo(getMapOptions(isMobile), [isMobile])`

That's premature optimization as is a plain object. Have a look at the article from Kent I've seen yesterday https://kentcdodds.com/blog/usememo-and-usecallback

Please, let's focus on the PR itself, I was just lazy to strip down that example more, wanted to show how that hook can be used :)

**Comment by JustFly1984** [comment]
useLoadScript should accept second argument `[appConfig.googleMapsApi, user.language]`

I have specific case where I want maps language to switch to correct language, when user changes the language in interface. This autodetection language feature is good enough, until you sit in some hotel in vietnam and try to use maps in vietnamese, cos it detects your browser. :-/

**Comment by FredyC** [comment]
> useLoadScript should accept second argument `[appConfig.googleMapsApi, user.language]`

I don't follow, why would be a benefit of the second argument? With component, it's also part of single props object.

File: packages/react-google-maps-api/src/LoadScriptNext.tsx [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+import { LoadScriptProps, DefaultLoadingElement } from './LoadScript'
+import { useLoadScript } from './useLoadScript'
+
+export interface LoadScriptNextProps extends LoadScriptProps {
+  loadingElement: React.ReactElement
+  children: React.ReactElement
+}
+
+export const LoadScriptNext = React.memo<LoadScriptNextProps>(
+  ({ loadingElement = <DefaultLoadingElement />, children, ...props }) => {
```
Comment (JustFly1984) [inline packages/react-google-maps-api/src/LoadScriptNext.tsx]: Please don't use default values for props on destructuring and use defaultProps instead. performance requirements.

File: packages/react-google-maps-api/src/LoadScriptNext.tsx [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+import { LoadScriptProps, DefaultLoadingElement } from './LoadScript'
+import { useLoadScript } from './useLoadScript'
+
+export interface LoadScriptNextProps extends LoadScriptProps {
+  loadingElement: React.ReactElement
+  children: React.ReactElement
+}
+
+export const LoadScriptNext = React.memo<LoadScriptNextProps>(
```
Comment (JustFly1984) [inline packages/react-google-maps-api/src/LoadScriptNext.tsx]: Please declare a component before wrapping it to React.memo. performance reasons.

Comment (FredyC) [reply packages/react-google-maps-api/src/LoadScriptNext.tsx]: Huh? Never head that one before. How is more performant if it's separated?

**Comment by FredyC** [comment]
> Please use named functions instead of arrow functions. better for debug, and `function` declarations hoist.

I already did that. Or did I missed a spot? :)

File: packages/react-google-maps-api/src/LoadScriptNext.tsx (hunk 5 above)
Comment (JustFly1984) [reply packages/react-google-maps-api/src/LoadScriptNext.tsx]: The correct way to declare a component as function first:
```jsx
function MyComponent () {
  return <JSX />
}
```

then you export it wrapped in memo:
```jsx
export default React.memo(MyComponent)
```

otherwise it ruins React debug with <undefined /> components.

Comment (JustFly1984) [reply packages/react-google-maps-api/src/LoadScriptNext.tsx]: React does check for Component.displayName, and React.memo does not provide it for us.

Comment (FredyC) [reply packages/react-google-maps-api/src/LoadScriptNext.tsx]: So you mean "debugging reasons". You got me fairly confused how that can be related to performance 😆

File: packages/react-google-maps-api/src/LoadScriptNext.tsx (hunk 4 above)
Comment (FredyC) [reply packages/react-google-maps-api/src/LoadScriptNext.tsx]: How do you handle that with TypeScript? I mean if I do `function LoadScriptNext` it does not have `defaultProps` and I would have to `ts-ignore` it which is ugly. The only way I know is to keep `const` which will keep the name as well, but allow proper typing.

```tsx
const LoadScriptNext: React.FC<LoadScriptNextProps> = ({ loadingElement, children, ...props }) => {
  const { isLoaded } = useLoadScript(props)
  return isLoaded ? children : loadingElement
}

LoadScriptNext.defaultProps = {
  loadingElement: <DefaultLoadingElement />
}

export default React.memo(LoadScriptNext)

```

Comment (FredyC) [reply packages/react-google-maps-api/src/LoadScriptNext.tsx]: Another approach would be like that, but I think const typing is much easier and approachable.

```ts
const defaultLoadingElement = <DefaultLoadingElement />

function LoadScriptNext({ loadingElement, children, ...props }: LoadScriptNextProps) {
  const { isLoaded } = useLoadScript(props)
  return isLoaded ? children : loadingElement || defaultLoadingElement
}

export default React.memo(LoadScriptNext)
```

File: packages/react-google-maps-api/src/LoadScriptNext.tsx (hunk 5 above)
Comment (JustFly1984) [reply packages/react-google-maps-api/src/LoadScriptNext.tsx]: sorry)

File: packages/react-google-maps-api/src/LoadScriptNext.tsx (hunk 4 above)
Comment (JustFly1984) [reply packages/react-google-maps-api/src/LoadScriptNext.tsx]:
```jsx
function LoadScriptNext(props: LoadScriptNextProps) {
  const { isLoaded } = useLoadScript(props)

  return isLoaded ? props.children : props.loadingElement
}

LoadScriptNext.defaultProps = {
  loadingElement: <DefaultLoadingElement />
}

export default React.memo(LoadScriptNext)
```

File: packages/react-google-maps-api/src/LoadScript.tsx [hunk 6]
```diff
@@ -26,15 +26,15 @@ export interface LoadScriptProps {
   preventGoogleFontsLoading?: boolean;
 }
 
-const DefaultLoadingElement = () => (
-  <div>{`Loading...`}</div>
-)
+export const DefaultLoadingElement = () => <div>{`Loading...`}</div>
```
Comment (JustFly1984) [inline packages/react-google-maps-api/src/LoadScript.tsx]: () =>

File: packages/react-google-maps-api/src/useLoadScript.ts [hunk 7]
```diff
@@ -0,0 +1,132 @@
+import * as React from 'react'
+
+import { injectScript } from './utils/injectscript'
+import { preventGoogleFonts } from './utils/prevent-google-fonts'
+
+import { isBrowser } from './utils/isbrowser'
+import { LoadScriptProps, defaultLoadScriptProps } from './LoadScript'
+
+export const useLoadScript = ({
```
Comment (JustFly1984) [inline packages/react-google-maps-api/src/useLoadScript.ts]: () =>

Comment (FredyC) [reply packages/react-google-maps-api/src/useLoadScript.ts]: Ah, this is the shame. Babel is actually handling that more properly and transpiles to named function. Sadly TypeScript doesn't bother, so yea, I will change that as well.

**Comment by FredyC** [comment]
Ok, I did a couple of changes, hopefully, it's better now :)

File: packages/react-google-maps-api/src/LoadScriptNext.tsx (hunk 4 above)
Comment (FredyC) [reply packages/react-google-maps-api/src/LoadScriptNext.tsx]: And what's the benefit of using `defaultProps` really? It did make sense for class components, but here it feels rather obsolete.

Comment (JustFly1984) [reply packages/react-google-maps-api/src/LoadScriptNext.tsx]: defaultProps does not get redeclared in compare with destructuring default values

Comment (JustFly1984) [reply packages/react-google-maps-api/src/LoadScriptNext.tsx]: basic types is pretty much safe, but objects, arrays and functions as default values recreated on each function call :-/

Comment (FredyC) [reply packages/react-google-maps-api/src/LoadScriptNext.tsx]: Sure, if you do `const defaultLoadingElement = <DefaultLoadingElement />` and referenced that, it's essentially the same thing, just bit more manual.

**Comment by FredyC** [comment]
@JustFly1984 Why that last commit? TypeScript will infer that return type on its own, doesn't it?

File: packages/react-google-maps-api/src/LoadScriptNext.tsx (hunk 4 above)
Comment (JustFly1984) [reply packages/react-google-maps-api/src/LoadScriptNext.tsx]: Component.defaultProps is not deprecated, and it is a recommended way to set default props :-)

Comment (FredyC) [reply packages/react-google-maps-api/src/LoadScriptNext.tsx]: Recommended? Would love to see docs link explaining that? :) 

image

**Comment by FredyC** [comment]
I just noticed open PR #137 and included that change in this PR as well.

**Comment by FredyC** [comment]
Rebased to master and ready for review :)

File: packages/react-google-maps-api/src/useLoadScript.ts [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+    prevLibraries.current = libraries
+  }, [libraries])
+
+  function makeUrl() {
+    const params = []
+
+    if (googleMapsApiKey) {
+      params.push(`key=${googleMapsApiKey}`)
+    } else if (googleMapsClientId) {
+      params.push(`client=${googleMapsClientId}`)
```
Comment (JustFly1984) [inline packages/react-google-maps-api/src/useLoadScript.ts]: if there is googleMapsClientId, no need to set googleMapsApiKey, and overwise, so basically googleMapsClientId should be first in if/else statement.

File: packages/react-google-maps-api/src/useLoadScript.ts [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+import { isBrowser } from './utils/isbrowser'
+import { defaultLoadScriptProps } from './LoadScript'
+import invariant from 'invariant';
+
+export interface UseLoadScriptOptions {
+  googleMapsApiKey?: string;
+  googleMapsClientId?: string;
+  id?: string;
+  version?: string;
+  language?: string;
```
Comment (JustFly1984) [inline packages/react-google-maps-api/src/useLoadScript.ts]: How do I change language in runtime?

File: packages/react-google-maps-api/src/useLoadScript.ts (hunk 8 above)
Comment (FredyC) [reply packages/react-google-maps-api/src/useLoadScript.ts]: I don't follow why the order matters, it's either API key or Client ID.

File: packages/react-google-maps-api/src/useLoadScript.ts (hunk 9 above)
Comment (FredyC) [reply packages/react-google-maps-api/src/useLoadScript.ts]: Just pass a new value to the hook? It will assemble new url and effect will re-rerun.

File: packages/react-google-maps-api/src/useLoadScript.ts (hunk 8 above)
Comment (JustFly1984) [reply packages/react-google-maps-api/src/useLoadScript.ts]: well, both options is optional, but if both options will be entered, clientId should win

Comment (FredyC) [reply packages/react-google-maps-api/src/useLoadScript.ts]: In the docs I only found this statement ...

> Note: If you were previously using an API key for authentication and are switching to using a client ID, you must remove the key parameter before loading the API. The API will fail to load if both a client ID and an API key are included.

It doesn't explicitly say that "client" is somehow "better". I think it should actually throw invariant to prevent surprising behavior. Someone might expect that key would be winning here. I will tweak that tomorrow.

File: packages/react-google-maps-api/src/useLoadScript.ts (hunk 9 above)
Comment (FredyC) [reply packages/react-google-maps-api/src/useLoadScript.ts]: Btw, in reality, changing language is not something the user does very often. For a fully localized app, it's usually better to just do a full reload to make sure everything is correctly loaded. It's surely a cool effect when it's changed in runtime, but not worth the hassle imo.

Comment (JustFly1984) [reply packages/react-google-maps-api/src/useLoadScript.ts]: in reality this is a part of my own app functionality, and it is important for my own project. sadly google doesn't provide an easy api to switch locale in runtime at all. I had this issue with original react-google-maps, and this functionality is one of main reasons why I started this project at first place.

Comment (FredyC) [reply packages/react-google-maps-api/src/useLoadScript.ts]: Ok, was just saying, but it should work nonetheless. You need to kinda try it, that's the purpose of this Next approach to see if it can work without that messy cleanup.

**Comment by FredyC** [comment]
Ok, next batch of changes. I've consolidated that url making function so it's shared and took care of both states when API key and ClientID or none of those are specified.

I've also documented these changes slightly. For some reason the styleguidist needs to have TSX extension for a hook otherwise it doesn't bother with loading it.

I wanted to document the long pending `useGoogleMap` as well, but styleguidist has a problem with the name of file `map-context` and won't load it even if renamed to TSX.

File: packages/react-google-maps-api/src/LoadScript.tsx [hunk 10]
```diff
@@ -179,27 +174,9 @@ class LoadScript extends React.PureComponent<LoadScriptProps, LoadScriptState> {
...
-    }
-
-    if (this.props.libraries && this.props.libraries.length) {
-      params.push(`&libraries=${this.props.libraries.join(",")}`)
-    }
-
     const injectScriptOptions = {
-      id: this.props.id,
-      url: `https://maps.googleapis.com/maps/api/js?${params.join('&')}`
+      id: this.props.id!,
```
Comment (JustFly1984) [inline packages/react-google-maps-api/src/LoadScript.tsx]: why `!` ?

Comment (FredyC) [reply packages/react-google-maps-api/src/LoadScript.tsx]: Because of stupid class & defaultProps :) TypeScript is not able to see that default value is assigned so it will never be undefined.

Comment (FredyC) [reply packages/react-google-maps-api/src/LoadScript.tsx]: For the reference, it was required a prop before, but since it has default, there is no point in requiring it.

Comment (JustFly1984) [reply packages/react-google-maps-api/src/LoadScript.tsx]: @FredyC 
I have issues with running gatsby-example locally after yarn workspaces merge, can you try to run it, you will see what I mean.
